File: ruby/test/test_scanner.rb
# frozen_string_literal: true

require "minitest/autorun"
require "tmpdir"
require_relative "../lib/envdoctor/scanner"

class ScannerTest < Minitest::Test
  def test_detects_usage_and_ignores_comments
    src = <<~RUBY
      # ENV["COMMENTED"]
      db = ENV["DB_URL"]
      port = ENV.fetch("PORT")
      =begin
      ENV["BLOCK_IGNORED"]
      =end
      user = ENV['DB_USER']
    RUBY
    used = Envdoctor::Scanner.scan_source("app.rb", src)
    assert_equal %w[DB_URL DB_USER PORT], used.keys.sort
    refute used.key?("COMMENTED")
    refute used.key?("BLOCK_IGNORED")
  end

  def test_reconciles_missing_and_unused
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "DB_URL=x\nUNUSED_KEY=1\n")
      File.write(File.join(dir, "app.rb"), "ENV[\"DB_URL\"]\nENV[\"NEW_FLAG\"]\n")
      findings = Envdoctor::Scanner.scan(dir)
      errors = findings.select { |f| f.severity == "error" }.map(&:name)
      warnings = findings.select { |f| f.severity == "warning" }.map(&:name)
      assert_includes errors, "NEW_FLAG"
      assert_includes warnings, "UNUSED_KEY"
      refute_includes errors, "DB_URL"
      refute_includes warnings, "DB_URL"
    end
  end

  def test_detects_duplicate_keys
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "DUP=a\nSOLO=1\nDUP=b\n")
      File.write(File.join(dir, "app.rb"), "ENV[\"DUP\"]\nENV[\"SOLO\"]\n")
      findings = Envdoctor::Scanner.scan(dir)
      dup = findings.find { |f| f.rule == "duplicates" }
      refute_nil dup
      assert_equal "DUP", dup.name
      assert_equal "error", dup.severity
      assert_equal "defined 2 times in the same file (lines 1, 3)", dup.message
      # single-definition key not reported as duplicate
      refute(findings.any? { |f| f.rule == "duplicates" && f.name == "SOLO" })
      # duplicated-but-used var must NOT also be flagged unused/undefined
      refute(findings.any? { |f| f.name == "DUP" && %w[unused undefined-in-source].include?(f.rule) })
    end
  end

  def test_detects_public_prefix_secrets
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"),
                 "NEXT_PUBLIC_API_KEY=x\nVITE_SECRET=y\nPUBLIC_URL=z\nAPI_KEY=w\nPUBLIC_KEY=k\n")
      findings = Envdoctor::Scanner.scan(dir)
      flagged = findings.select { |f| f.rule == "public-prefix" }.map(&:name)
      assert_includes flagged, "NEXT_PUBLIC_API_KEY"
      assert_includes flagged, "VITE_SECRET"
      refute_includes flagged, "PUBLIC_URL"
      refute_includes flagged, "API_KEY"
      refute_includes flagged, "PUBLIC_KEY"
      flagged.each do |name|
        f = findings.find { |x| x.rule == "public-prefix" && x.name == name }
        assert_equal "error", f.severity
      end
    end
  end
end

File: python/tests/test_scanner.py
from pathlib import Path

from envdoctor.scanner import scan, scan_source_file


def _write(tmp_path: Path, name: str, content: str) -> Path:
    p = tmp_path / name
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(content)
    return p


def test_detects_all_usage_forms(tmp_path):
    src = _write(
        tmp_path,
        "app.py",
        "import os\n"
        "from os import environ\n"
        "# os.getenv('COMMENTED')\n"
        "a = os.getenv('DB_URL')\n"
        "b = os.environ.get('PORT')\n"
        "c = os.environ['API_KEY']\n"
        "d = environ.get('HOST')\n"
        "e = environ['DB_USER']\n"
        '"""docstring os.getenv("DOC_IGNORED")"""\n',
    )
    used = scan_source_file(src)
    assert set(used) == {"DB_URL", "PORT", "API_KEY", "HOST", "DB_USER"}
    assert "COMMENTED" not in used
    assert "DOC_IGNORED" not in used


def test_missing_and_unused(tmp_path):
    _write(tmp_path, ".env", "DB_URL=postgres://x\nUNUSED_KEY=1\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DB_URL')\nos.getenv('NEW_FLAG')\n")

    result = scan(tmp_path)
    errors = {f.name for f in result.errors}
    warnings = {f.name for f in result.warnings}

    assert "NEW_FLAG" in errors  # used but not defined
    assert "UNUSED_KEY" in warnings  # defined but not used
    assert "DB_URL" not in errors and "DB_URL" not in warnings  # reconciled


def test_duplicates(tmp_path):
    _write(tmp_path, ".env", "DB_URL=a\nDB_URL=b\nSOLO=1\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DB_URL')\nos.getenv('SOLO')\n")

    result = scan(tmp_path)
    dups = [f for f in result.findings if f.rule == "duplicates"]
    assert len(dups) == 1
    assert dups[0].name == "DB_URL"
    assert dups[0].severity == "error"
    assert "lines 1, 2" in dups[0].message
    # Single-definition key is not reported as a duplicate.
    assert "SOLO" not in {f.name for f in dups}
    # First occurrence still reconciles: DB_URL is not undefined/unused.
    assert "DB_URL" not in {f.name for f in result.warnings}


def test_public_prefix(tmp_path):
    _write(
        tmp_path,
        ".env",
        "NEXT_PUBLIC_API_KEY=x\nPUBLIC_URL=x\nAPI_KEY=x\n",
    )
    result = scan(tmp_path)
    pp = [f for f in result.findings if f.rule == "public-prefix"]
    names = {f.name for f in pp}
    assert names == {"NEXT_PUBLIC_API_KEY"}
    assert pp[0].severity == "error"
    assert "PUBLIC_URL" not in names  # public prefix but not secret-like
    assert "API_KEY" not in names  # secret-like but no public prefix


def test_clean_project_has_no_findings(tmp_path):
    _write(tmp_path, ".env", "DB_URL=x\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DB_URL')\n")
    result = scan(tmp_path)
    assert result.findings == []
